### ISSUE TITLE
hangs on returning HIT

### ISSUE THREAD

**Issue by Nitradamus** [jiangwenyuan/nuster#30]
I'm trying to play around with Nuster in front of nginx/varnish and I've stumbled upon a issue where if a key is set, the next hit doesn't return any output back to the client.

The first  request goes all good, but the second one (which should return the cached content) just hangs and doesn't output anything resulting in a non-working site.

It seems to only be when using Varnish as backend, and also sometimes on static files (a .woff file for example after a ctrl+f5)

I was wondering if I'm doing something wrong or this is a bug? This is preventing me from considering it for general use so I hope you can point me in the right direction :)

Nuster config:
 ```
global
    log 127.0.0.1 local0
    nuster cache on data-size 250m uri /_nuster
    maxconn 1000000
    daemon
    nbproc 2
    tune.maxaccept -1
    tune.ssl.default-dh-param 2048
    tune.h2.max-concurrent-streams 1000
    user nginx
    group nginx
    debug
defaults
    retries 3
    maxconn 1000000
    option redispatch
    option dontlognull
    option http-keep-alive
    timeout client  300s
    timeout connect 300s
    timeout server  300s
    http-reuse always
    mode http
    log global
frontend frontend
    bind *:80
    bind *:443 ssl crt /etc/ssl/certs/cert.pem alpn h2,http/1.1
    #redirect http to https
    redirect scheme https code 301 if !{ ssl_fc }
    rspadd Strict-Transport-Security:\ max-age=31536000;\ includeSubDomains;\ preload
    rspadd X-Frame-Options:\ DENY

    acl is_get method GET HEAD
    use_backend cache unless !is_get 

    default_backend nocache
backend cache
    nuster cache on
    nuster rule all ttl 10s
    server varnish01 /var/run/varnish.sock
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-response set-header X-Nuster-Cache ON
backend nocache
    server nginx01 /var/run/nginx.sock
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    http-response set-header X-Nuster-Cache OFF
 ```

the output of Nuster:

First request:

 ```
Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result FAILED
Total: 3 (2 usable), will use epoll.

Available filters :
	[SPOE] spoe
	[COMP] compression
	[TRACE] trace
Using epoll() as the polling mechanism.
[CACHE] on, data_size=262144000
00000000:frontend.accept(0005)=0009 from [1.1.1.1:60308] ALPN=h2
00000000:frontend.clireq[0009:ffffffff]: GET / HTTP/1.1
00000000:frontend.clihdr[0009:ffffffff]: user-agent: curl/7.61.0
00000000:frontend.clihdr[0009:ffffffff]: accept: */*
00000000:frontend.clihdr[0009:ffffffff]: host: www.domain.com
[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: GET.HTTPS.www.domain.com./.
[CACHE] Checking key existence: NOT EXIST
[CACHE] [REQ] Checking if rule pass: PASS
00000000:cache.srvrep[0009:000a]: HTTP/1.1 200 OK
00000000:cache.srvhdr[0009:000a]: Server: nginx/1.14.0
00000000:cache.srvhdr[0009:000a]: Date: Wed, 31 Oct 2018 12:54:18 GMT
00000000:cache.srvhdr[0009:000a]: Content-Type: text/html; charset=UTF-8
00000000:cache.srvhdr[0009:000a]: X-Powered-By: PHP/7.2.11
00000000:cache.srvhdr[0009:000a]: X-Pingback: https://www.domain.com/cms/xmlrpc.php
00000000:cache.srvhdr[0009:000a]: Link: <https://www.domain.com/>; rel=shortlink
00000000:cache.srvhdr[0009:000a]: X-Varnish: 393226
00000000:cache.srvhdr[0009:000a]: Age: 0
00000000:cache.srvhdr[0009:000a]: Via: 1.1 varnish (Varnish/6.0)
00000000:cache.srvhdr[0009:000a]: Accept-Ranges: bytes
00000000:cache.srvhdr[0009:000a]: Content-Length: 25245
00000000:cache.srvhdr[0009:000a]: Connection: keep-alive
[CACHE] [RES] Checking status code: PASS
[CACHE] To create
00000001:frontend.clicls[0009:ffffffff]
00000001:frontend.closed[0009:ffffffff]
 ```
Second request, should be cached now:

 ```[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: GET.HTTPS.www.domain.com./.
[CACHE] Checking key existence: EXIST
[CACHE] Hit
 ```

and just sits there and does nothing anymore until killed. 

I hope you can point me in the right direction, awesome project!

**Comment by Nitradamus** [comment]
Not sure if this helps in any way, but running with strace shows the following on the second hit:

```[CACHE] Hit
)    = 18
epoll_wait(3, [], 200, 0)               = 0
write(9, "\27\3\3\0000\266\20\257\10&\362\0\2264\217\256\260\320\35K\225'\336\202f\254\200\215\342\375\f\250"..., 53) = 53
epoll_wait(3, [], 200, 0)               = 0
epoll_wait(3, [{EPOLLIN, {u32=9, u64=9}}], 200, 1000) = 1
read(9, "\27\3\3\0!", 5)                = 5
read(9, "\344m\275Q\17\31\367\321\232\356\307\305\323\313\251d\3335T\353\212\235h\347\240\311c\244\215\301(*"..., 33) = 33
read(9, 0x19c2013, 5)                   = -1 EAGAIN (Resource temporarily unavailable)
```

**Comment by Nitradamus** [comment]
hi @jiangwenyuan, thanks for the help thus far. I made sure Varnish and/or nginx wasn't the problem, both backends give me the same result.

I can recreate it by using the following php code on index.php

<?php phpinfo(); ?>

only requesting the headers with curl -I https://domain.com works fine:

```Using epoll() as the polling mechanism.
[CACHE] on, data_size=262144000
00000000:frontend.accept(0005)=0009 from [109.109.102.102:40124] ALPN=h2
00000000:frontend.clireq[0009:ffffffff]: HEAD / HTTP/1.1
00000000:frontend.clihdr[0009:ffffffff]: user-agent: curl/7.61.0
00000000:frontend.clihdr[0009:ffffffff]: accept: */*
00000000:frontend.clihdr[0009:ffffffff]: host: domain.com
[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: HEAD.HTTPS.domain.com./.
[CACHE] Checking key existence: NOT EXIST
[CACHE] [REQ] Checking if rule pass: PASS
00000000:cache.srvrep[0009:000a]: HTTP/1.1 200 OK
00000000:cache.srvhdr[0009:000a]: Date: Fri, 02 Nov 2018 10:16:38 GMT
00000000:cache.srvhdr[0009:000a]: Content-Type: text/html; charset=UTF-8
00000000:cache.srvhdr[0009:000a]: Age: 101
00000000:cache.srvhdr[0009:000a]: X-Cache: HIT
00000000:cache.srvhdr[0009:000a]: X-Cache-Hits: 2
00000000:cache.srvhdr[0009:000a]: Accept-Ranges: bytes
00000000:cache.srvhdr[0009:000a]: Connection: keep-alive
[CACHE] [RES] Checking status code: PASS
[CACHE] To create
00000001:frontend.clicls[0009:ffffffff]
00000001:frontend.closed[0009:ffffffff]
00000002:frontend.accept(0005)=0009 from [109.109.102.102:40150] ALPN=h2
00000002:frontend.clireq[0009:ffffffff]: HEAD / HTTP/1.1
00000002:frontend.clihdr[0009:ffffffff]: user-agent: curl/7.61.0
00000002:frontend.clihdr[0009:ffffffff]: accept: */*
00000002:frontend.clihdr[0009:ffffffff]: host: domain.com
[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: HEAD.HTTPS.domain.com./.
[CACHE] Checking key existence: EXIST
[CACHE] Hit
00000002:cache.srvrep[0009:ffffffff]: HTTP/1.1 200 OK
00000002:cache.srvhdr[0009:ffffffff]: Date: Fri, 02 Nov 2018 10:16:38 GMT
00000002:cache.srvhdr[0009:ffffffff]: Content-Type: text/html; charset=UTF-8
00000002:cache.srvhdr[0009:ffffffff]: Age: 101
00000002:cache.srvhdr[0009:ffffffff]: X-Cache: HIT
00000002:cache.srvhdr[0009:ffffffff]: X-Cache-Hits: 2
00000002:cache.srvhdr[0009:ffffffff]: Accept-Ranges: bytes
00000002:cache.srvhdr[0009:ffffffff]: X-Nuster-Cache: ON
00000002:cache.srvhdr[0009:ffffffff]: Strict-Transport-Security: max-age=31536000; includeSubDomains; preload
00000002:cache.srvhdr[0009:ffffffff]: X-Frame-Options: DENY
00000003:frontend.clicls[0009:ffffffff]
00000003:frontend.closed[0009:ffffffff]
```

normal GET:

```Using epoll() as the polling mechanism.
[CACHE] on, data_size=262144000
00000000:frontend.accept(0005)=0009 from [109.109.102.102:40396] ALPN=h2
00000000:frontend.clireq[0009:ffffffff]: GET / HTTP/1.1
00000000:frontend.clihdr[0009:ffffffff]: user-agent: curl/7.61.0
00000000:frontend.clihdr[0009:ffffffff]: accept: */*
00000000:frontend.clihdr[0009:ffffffff]: host: domain.com
[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: GET.HTTPS.domain.com./.
[CACHE] Checking key existence: NOT EXIST
[CACHE] [REQ] Checking if rule pass: PASS
00000000:cache.srvrep[0009:000a]: HTTP/1.1 200 OK
00000000:cache.srvhdr[0009:000a]: Date: Fri, 02 Nov 2018 10:16:38 GMT
00000000:cache.srvhdr[0009:000a]: Content-Type: text/html; charset=UTF-8
00000000:cache.srvhdr[0009:000a]: Age: 263
00000000:cache.srvhdr[0009:000a]: X-Cache: HIT
00000000:cache.srvhdr[0009:000a]: X-Cache-Hits: 4
00000000:cache.srvhdr[0009:000a]: Accept-Ranges: bytes
00000000:cache.srvhdr[0009:000a]: Content-Length: 88104
00000000:cache.srvhdr[0009:000a]: Connection: keep-alive
[CACHE] [RES] Checking status code: PASS
[CACHE] To create
00000001:frontend.clicls[0009:ffffffff]
00000001:frontend.closed[0009:ffffffff]
00000002:frontend.accept(0005)=0009 from [109.109.102.102:40398] ALPN=h2
00000002:frontend.clireq[0009:ffffffff]: GET / HTTP/1.1
00000002:frontend.clihdr[0009:ffffffff]: user-agent: curl/7.61.0
00000002:frontend.clihdr[0009:ffffffff]: accept: */*
00000002:frontend.clihdr[0009:ffffffff]: host: domain.com
[CACHE] Checking rule: all
[CACHE] Calculate key: method.scheme.host.uri.
[CACHE] Got key: GET.HTTPS.domain.com./.
[CACHE] Checking key existence: EXIST
[CACHE] Hit
```

^ Freezes at HIT

I tried changing nginx/varnish config, only using nginx as backend etc. but I guess the request gets fetched just fine, it's the cache retrieving part in Nuster that's somehow not working, but why does it work for a HEAD and not a GET?

Any idea what could be wrong? I'm out of options :)

**Comment by Nitradamus** [comment]
Weirdly enough, everything is fine when I just use the following code instead of phpinfo():

`<?php echo 'hi' ?>`

Could it be because of the size of the result?

**Comment by jiangwenyuan** [comment]
Hi, finally found out the cause

```
   rspadd Strict-Transport-Security:\ max-age=31536000;\ includeSubDomains;\ preload
   rspadd X-Frame-Options:\ DENY
   ...
   http-response set-header X-Nuster-Cache ON
```
If you comment out first two lines, there won't be any problem.

The cause is that haproxy uses a buf with default size 16K, if an http response(headers) is larger
than this size, haproxy will return HTTP 502.

In order to save memory, nuster stores first part or response body along with headers, and if you try to append long headers, there will be no room for that.

You can add these headers in nginx as a tentative workaround.

**Comment by jiangwenyuan** [comment]
Hi @Nitradamus , fixed in 49cba5788a330cec415b4e7def90a2f90e451b51 , could you please take a look? Appreciate it:)

**Comment by Nitradamus** [comment]
> Hi @Nitradamus , fixed in [49cba57](https://github.com/jiangwenyuan/nuster/commit/49cba5788a330cec415b4e7def90a2f90e451b51) , could you please take a look? Appreciate it:)

Thanks for this! I tried it out and it seemed to fix the problem at first however for me it creates a new one when the content-encoding is br or gzip. Now it spends a lot of time in '[CACHE] To create' but it already has the content from the backend by then?

Reverting back to the original code and commenting out the rspadd values like you said fixes everything..

Any idea? :)

**Comment by jiangwenyuan** [comment]
Hi @Nitradamus Thanks, fixed in [fixes/30 ](https://github.com/jiangwenyuan/nuster/tree/fixes/30)

I could manage to have some time to do a thorough test this weekend, it would be a great help if you can verify on your side, thanks:)

> Besides that: is there a way to define different caching keys for different encoding methods? i.e. now it returns br encoded content even if the browser doesn't support it if was cached before in a browser that does.

Currently nuster does not support `Vary`, you can specify a custom key like `   nuster rule all key method.scheme.host.uri.header_Accept-Encoding ttl 0` as a workaround.

Cheers

**Comment by Nitradamus** [comment]
Thanks for the update! It seems it's working fine now after testing it shortly. I will let you know if anything comes up :)

I'll try the suggested custom key method to deal with the content encoding!

**Comment by Nitradamus** [comment]
> Currently nuster does not support `Vary`, you can specify a custom key like ` nuster rule all key method.scheme.host.uri.header_Accept-Encoding ttl 0` as a workaround.
> 
> Cheers

Hi @jiangwenyuan, this method doesn't work as it should as it seems to only take the first setting of the Accept-Encoding header:

00000048:frontend.clihdr[0009:ffffffff]: accept-encoding: gzip, deflate, br

[CACHE] Calculate key: method.scheme.host.uri.header_Accept-Encoding.
[CACHE] Got key: GET.HTTPS.www.domain.com./folder/.gzip.

Any idea ? :)

**Comment by jiangwenyuan** [comment]
Ah, right, only the first occurrence of the value is used in case of comma-separated header. This should be improved. sorry about that

**Comment by Nitradamus** [comment]
> Ah, right, only the first occurrence of the value is used in case of comma-separated header. This should be improved. sorry about that

No need to be sorry, but I guess  this has to be fixed within Nuster? If so, no pressure ;)

**Comment by jiangwenyuan** [comment]
Yes, sure, I'm working on `vary` and other stuff (very limited spare time:(, contributors wanted:)
I will merge the fix into master and close this issue, feel free to fire issue any time:)
thanks

**Comment by Nitradamus** [comment]
> Yes, sure, I'm working on `vary` and other stuff (very limited spare time:(, contributors wanted:)
> I will merge the fix into master and close this issue, feel free to fire issue any time:)
> thanks

Hello @jiangwenyuan, I was wondering if this (vary) was supposed to be fixed yet? I still get the same issues with different content-encoding values between clients and not sure how to get around this as of yet :(

**Comment by jiangwenyuan** [comment]
Hi, unfortunately not yet:( I was busy with migration to HAProxy v1.9 and planned to work out a full support of `vary` in v1.9.

I'll make a workaround in v1.8, since there's a critical fix in HAProxy v1.8 recently, and I'm merging that:)

**Comment by Nitradamus** [comment]
> Hi, unfortunately not yet:( I was busy with migration to HAProxy v1.9 and planned to work out a full support of `vary` in v1.9.
> 
> I'll make a workaround in v1.8, since there's a critical fix in HAProxy v1.8 recently, and I'm merging that:)

Awesome! Can't wait to try it out! Thanks a lot :)

**Comment by jiangwenyuan** [comment]
@Nitradamus See the newest release, still need to use `header_Accept-Encoding` though.
